[Javiery3889] iP

Javiery3889 opened the pull request:
# Blu

> "Why waste time say lot word when few word do trick?"

Tired of using pretty GUIs to manage your tasks? Well fear not, Blu is a CLI task management app that aims to provide:
- ~fast~ **REALLY FAST** CLI
- Intuitive commands
- Informative messages

## Setup Guide 🔧
1. Download the JAR file [here](https://github.com/Javiery3889/ip/releases/tag/A-Jar)
2. Open a terminal and navigate to the location where you saved the above file.
3. Launch the JAR using the `java -jar blu.jar`

## Developer Guide
Below is the entrypoint of the application is `blu.java`.
```java
public static void main(String[] args) {
    new Blu().run();
}
```

## Future Improvements
- [x] Use Checkstyle plugin in Gradle
- [ ] Handle multiple forms of datetime formats input
- [ ] OpenAI API 👀

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Javiery3889 committed:
Rename Bot. Added Greet and Exit messages.

Javiery3889 committed:
Added Echo functionality

Javiery3889 committed:
Added List, Add To List functionality

- Refactored print method in Chatbot.java to accept String arrays
- Created Task.java to represent each user created task

Javiery3889 committed:
Refactor code for more OOP

- Created InputHandler.java to separate input handling from Blu.java
- Operations on list moved to TaskList.java
- Added toString method for Task to print out task's title

Javiery3889 committed:
Added Mark/Unmark task functionality

- Change getTask in TaskList.java to use 1-based indexing

Javiery3889 committed:
Fixed list numbering bug

Javiery3889 committed:
Added tracking of three subtypes of tasks

- Refactored InputHandler to use switch/case statements
- Added method to retrieve number of tasks in TaskList.java
- Change addTask method in Chatbot.java to accept Task object instead

Javiery3889 committed:
Change runtest.sh for testing

Javiery3889 committed:
Added exceptions to handle errors

- Fixed bug in InputHandler.java where first word of deadline/event
  description was not included
- Created BluException class and three exception subclasses
  - IllegalCommand used when there are missing parameters
  - InvalidCommand used for commands that are not supported
  - IllegalParameter used when parameter does not satisfy given
    constraints
- runtest.sh now complies all java files in subdirectories

Javiery3889 committed:
Add delete task functionality

- Refactored InputHandler to call a method to validate number of command
  parameters

Javiery3889 committed:
Changed command parsing to use enums

Javiery3889 committed:
feat: Save and load tasks from file

Javiery3889 committed:
Merge branch 'branch-Level-7'

Javiery3889 committed:
feat: Parse datetime for deadline and event tasks

Javiery3889 committed:
Merge branch 'branch-Level-8'

Javiery3889 committed:
Refactor codebase to adhere OOP principles

Blu.java now creates new instance of Blu

Abstract each command to its respective Command class

Input and Output handling now performed only in UI.java

Create StorageException.java to handle exceptions during
task storage operations

Javiery3889 committed:
Merge branch 'A-MoreOOP'

Javiery3889 committed:
Divide classes into respective packages

Javiery3889 committed:
Merge branch 'A-Packages'

Javiery3889 committed:
Rename package to match coding standard

Javiery3889 committed:
Merge branch 'add-gradle-support'

Javiery3889 committed:
Add gradle support

Javiery3889 committed:
Move error handling to respective files

Previously storage and tasklist exceptions in each command
were handled in each Command subclass.

Storage and tasklist exceptions are now handled in their
respective .java file

Fix bug in UnmarkCommand.java where unmark command
was not correctly unmarking the task given

Fix bug in InputParser.java where switch statement
was not returning DeleteCommand class

Javiery3889 committed:
Add junit tests

Add J-Unit tests for TaskDecoder.java and TaskList.java

Fixed a bug in TaskDecoder where "==" operator was used instead
of ".equals()" for comparing strings

Javiery3889 committed:
Merge branch 'A-JUnit'

Javiery3889 committed:
Add javadocs

Javiery3889 committed:
Merge branch 'A-JavaDoc'

Javiery3889 committed:
Refactor code to meet style guide

Javiery3889 committed:
Merge branch 'A-CodingStandard'

Javiery3889 committed:
Add find command

Javiery3889 committed:
Merge branch 'Level-9'

Javiery3889 committed:
Refactor code to pass checkstyle

Also added JavaDocs for:
- Event.java
- Todo.java
- Deadline.java
- Blu.java

Javiery3889 committed:
Merge branch 'A-CheckStyle'

Javiery3889 committed:
Add basic GUI support

Javiery3889 committed:
Merge branch 'Level-10'

Javiery3889 committed:
Use Varargs for displaying messages

Javiery3889 committed:
Merge branch 'A-Varargs'

Xilef121 commented on src/main/java/Blu/command/DeadlineCommand.java:
Could consider having a general TaskCommand with 3 different constructors to create the 3 differenent tasks rather than having DeadlineCommand, EventCommand, TodoCommand.

Xilef121 commented on src/main/java/Blu/task/TaskList.java:
```diff
@@ -0,0 +1,140 @@
+     * @param searchString The substring to search for within each task of the task list.
+     * @return A list of task that contain the specified substring.
+     */
+    public List<Task> searhForTasksContaining(String searchString) {
```
I believe there is a typo in your method name

Javiery3889 replied:
Thank you for spotting this!

Xilef121 reviewed (commented):
Overall, the code looks good, abiding by the coding standard for the names of classes and methods. OOP is also done well with good abstraction. Javadocs are done well for public methods.

Javiery3889 committed:
Fixed bug where welcome message is not displayed

Upon startup, Blu does not display welcome message as expected.
This is because the method to display the welcome message is
never called in the application.

Create the welcome message in init() and displays it in start().
Change welcome message to include canonical path of the storage
file instead.

This allows users to locate the file path easily.

Javiery3889 committed:
Add assertions to non-empty command and file write

Program assumes command parse by input parser is always a
non-null value. It also assumes that the tasks will be
successfully saved to the file.

This can lead to unintended bugs caused by the developer,
as both cases are not supposed to be empty under any
circumstances.

Use assert statements to indicate a possible bug within
these two cases, which violates the assumption that both
commands and file write operations should be non-empty.

This is allow runtime to detect any bugs made by the developer,
allowing the program to terminate earlier safely.

Javiery3889 committed:
Refactor codebase to improve code quality

- Remove magic values in TaskDecoder.java
- Simplify boolean logic in Deadline.java and Event.java
- Minimise code duplication in InputParser.java, specifically
in handling parsing of integers and localdatetimes.

Javiery3889 committed:
Use Streams to print task list

Javiery3889 committed:
Merge pull request #1 from Javiery3889/branch-A-Assertions

Add assertions to non-empty command and file write

Javiery3889 committed:
Merge pull request #2 from Javiery3889/branch-A-CodeQuality

Refactor codebase to improve code quality

Javiery3889 committed:
Merge pull request #3 from Javiery3889/branch-A-Streams

Use Streams to print task list

Javiery3889 committed:
Set up CI

Javiery3889 committed:
Add option for user to choose data source

Javiery3889 committed:
Merge branch 'branch-BCD-Extension'

felixchanyy commented on src/main/java/Blu/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,47 @@
+    @Override
+    public String execute(TaskList taskList, Storage storage, UI ui) throws StorageException {
+        Deadline deadline = new Deadline(description, byDateTime);
+        taskList.addTask(deadline);
```
I like how verbs are used for methods

felixchanyy commented on src/main/java/Blu/command/EventCommand.java:
```diff
@@ -0,0 +1,50 @@
+ * Represents a command to add a new Event task with a specified description and event date/time.
+ */
+public class EventCommand extends Command {
+    private final String description;
```
I like how description, fromDateTime, and toDateTime convery their respective meanings and are single-valued attribute variables.

felixchanyy commented on src/main/java/Blu/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,45 @@
+     *
+     * @param taskIdx The index of the task to be unmarked.
+     */
+    public UnmarkCommand(int taskIdx) {
```
I like how UnmarkCommand conveys its purpose, a command related to unmarking a task.

felixchanyy commented on src/main/java/Blu/parser/InputParser.java:
```diff
@@ -0,0 +1,211 @@
+     * @return The command corresponding to the user input.
+     * @throws IllegalCommandException If the input cannot be parsed into a valid command.
+     */
+    public Command parseInput(String userInput) throws IllegalCommandException {
```
I like how parseInput follows a sensible order.

felixchanyy commented on src/main/java/Blu/storage/Storage.java:
```diff
@@ -0,0 +1,120 @@
+     * @return A TaskList containing the tasks loaded from the file.
+     * @throws StorageException If tasks cannot be read from the storage file.
+     */
+    public TaskList loadTasks() throws StorageException {
```
I like how loadTasks() is short in length yet descriptive of its functionality.

felixchanyy reviewed (commented):
I like how clean this code is.

Javiery3889 committed:
Change error messages text to red

Javiery3889 committed:
Add personalised goodbye message

Javiery3889 committed:
Fix bug where user message is not displayed

Javiery3889 committed:
Fix issue where messages were cut off

Enable resizing of GUI window

Javiery3889 committed:
Update user guide

Javiery3889 committed:
Fix GFMD alert rendering

Javiery3889 committed:
Fix GFMD formatting

Javiery3889 committed:
Fix indentation in published view

Javiery3889 committed:
Fix DateTime format error message

Javiery3889 committed:
Fix typos in view
